src/Microsoft.Health.Fhir.CosmosDb/Features/Storage/Queues/CosmosQueueClient.cs

```diff
@@ -331,6 +331,11 @@ public async Task<bool> PutJobHeartbeatAsync(JobInfo jobInfo, CancellationToken
             var job = jobs.Single();
             JobDefinitionWrapper item = job.MatchingJob.Single();
 
+            if (item.CancelRequested)
+            {
+                return true; // If the job is cancelled, the job data doesn't need to be updated.
```

Comment:
I wouldn't say this is a safe assumption to make. Even if the job is cancelled it may have data that needs to be updated to show what point the job got to before being cancelled.

Reply:
Good call - I don't think this will impact perf of the cancel call but will run a test tomorrow to make sure. If the `CancelRequested` flag is set, the cancellation job won't be trying to change the document at the same time.

The issue is the heartbeats, completed jobs, new jobs and the cancels all at the same time was causing the parallel threads to compete in their optimistic concurrency. I may have gone overboard trying to fix it. 😫